keps/sig-node/3322-add-maxRestartTimes-to-podSpec/README.md

```diff
@@ -0,0 +1,981 @@
+will be controlled by `maxRestartTimesOnFailure`, when a Pod hits the value+1, it will be marked as `Failed`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+When running with a pod, some system errors or program internal errors
+encountered, which causes the application restarting infinitely. This is
+unrecoverable and wasteful.
+
+### Story 2
+
+For batch Job specifically, now a Job provides a single backoffLimit for all derived pods,
+which may cause the entire Job failed and clean up all the resources because of one of the pods
+restarts for more than the `spec.backoffLimit`. To solve the problem, the Batch WG incubates
+a new KEP [Backoff Limit Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/issues/3850)
+which introduces a new field `backoffLimitPerIndex` to control the max restart times per Pod.
+
+It's actually what we want to implement here, by sinking the capacity of `backoffLimitPerIndex` to pod level,
+it helps to reduce the maintenance costs of the Job API.
+
+Note: currently, `backoffLimitPerIndex` only supports `restartPolicy=Never` initially, but as the KEP states, it will depend
+on this feature to support onFailure restartPolicy in the future, see [chapter](https://github.com/kubernetes/enhancements/blob/ab54c14de86af3bcc76b45a4c9194bc47f6c9a62/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md#support-backofflimitperindex-when-restartpolicyonfailure).
+This reminds me whether we should widen the _max_restart_times_ to other restart policies like `Never`, see discussion below.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Discussions
+
+The Batch WG incubates a new KEP [Backoff Limit Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/issues/3850),
+which introduces a new field `backoffLimitPerIndex`, then each pod of the Job will have an independent backOffLimit,
+but initially it only supports `restartPolicy=Never`, maybe we can widen the _max_restart_times_ to support other restart policies,
+like `Never`, then the Job API no longer needs to consider these logics anymore. Currently, it uses an annotation `batch.kubernetes.io/job-index-failure-count`
+to record the retires number, which looks not elegant.
+
+Pros:
+  * BackoffLimitPerIndex can reuse this functionality and no longer need to consider the restart times per index.
+  Specifically, it can avoid to use the annotation anymore, and works at a high level control by watching the pod status.
```

Comment:
OK, that's something beyond the KEP. Updated.

Reply:
I actually agree with Aldo.
It's not an implementation detail - it's a fundamental thing of "pod recreations". If we want to track something across pod recreations (which is the case for jobs), maxRestarts won't solve it on its own - but actually may help with it.

Reply:
> OK, the implementation details of BackoffLimitPerIndex is something beyond the KEP.

It's not currently stated in the Non-Goals. But I think doing so would be a mistake. Even if we don't have the full semantics for how the pod level restart limit would work with backoffLimit or backoffLimitPerIndex, we should have some idea of how they could work together. Otherwise we could end up in another scenario where features work independently, but when paired together are mostly useless or have very strange semantics that are hard for users to understand, and hard to maintain.

Reply:
> Otherwise we could end up in another scenario where features work independently, but when paired together are mostly useless or have very strange semantics that are hard for users to understand, and hard to maintain.

I like this analysis.